stdlib/public/core/StringWordView.swift

```diff
@@ -0,0 +1,120 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension String {
+  internal struct _WordView {
+    internal var _guts: _StringGuts
+
+    internal init(_ _guts: _StringGuts) {
+      self._guts = _guts
+    }
+  }
+}
+
+extension String._WordView: Collection {
+  internal struct Index {
+    @usableFromInline
+    internal var _encodedOffset: Int
+  }
+
+  typealias Element = Substring
+
+  var startIndex: Index {
+    Index(_encodedOffset: _guts.startIndex._encodedOffset)
+  }
+
+  var endIndex: Index {
+    Index(_encodedOffset: _guts.endIndex._encodedOffset)
+  }
+
+  func index(after i: Index) -> Index {
+    _internalInvariant(i < endIndex)
+
+    let nextOffset = _guts._opaqueNextWordIndex(startingAt: i._encodedOffset)
+    return Index(_encodedOffset: nextOffset)
+  }
+
+  subscript(position: Index) -> Element {
+    let indexAfter = index(after: position)
+
+    let l = String.Index(_encodedOffset: position._encodedOffset)
+    let u = String.Index(_encodedOffset: indexAfter._encodedOffset)
+
+    return String(_guts)[l ..< u]
+  }
+}
+
+extension String._WordView.Index: Equatable {
+  @inlinable
+  @inline(__always)
+  static func == (lhs: Self, rhs: Self) -> Bool {
+    lhs._encodedOffset == rhs._encodedOffset
+  }
+}
+
+extension String._WordView.Index: Comparable {
+  @inlinable
+  @inline(__always)
+  static func < (lhs: Self, rhs: Self) -> Bool {
+    lhs._encodedOffset < rhs._encodedOffset
+  }
+}
+
+extension String._WordView.Index {
+  internal init(_ i: String.Index) {
+    _encodedOffset = i._encodedOffset
+  }
+}
+
+extension String._WordView.Index: Sendable {}
+
+extension String._WordView: BidirectionalCollection {
+  func index(before i: Index) -> Index {
+    _internalInvariant(i > startIndex)
+
+    let previousOffset = _guts._opaquePreviousWordIndex(
+      endingAt: i._encodedOffset
+    )
+
+    return Index(_encodedOffset: previousOffset)
+  }
+}
+
+extension String._WordView: Sendable {}
+
+extension String {
+  // FIXME: Figure out what to do with this if/when we actually release it as
+  // public API.
+  // Should this be:
+  //    var words: WordView
+  // or perhaps
+  //    func words(_ level: ...) -> some BidirectionalCollection<Substring>
```

Comment:
☝️

Reply:
Using an opaque result type for this would make it impossible to integrate this new collection type into the existing String design in any meaningful sense.

E.g., it would mean that we'd be giving up on ever being able to add methods to convert between `String.Index` and the custom Index type of this collection (assuming that we actually want a custom index type for this thing, which I continue to be skeptical about). To me it seems like a reasonable expectation that the stdlib would provide a way to quickly find the index of a word that contains a particular character or scalar index.

The stdlib ought to be a cohesive library with well-integrated parts, not just a disjoint collection of independent components.

Reply:
> Using an opaque result type for this would make it impossible to integrate this new collection type into the existing String design in any meaningful sense.

Exactly

Reply:
> Exactly

What do you mean? Could you elaborate please?